Amaresh taking over for Linh – Fix Educator Task Submissions UI + Late Submission Detection Improvements

linh2020 opened the pull request:
# Description
<img width="625" height="439" alt="image" src="https://github.com/user-attachments/assets/cf1ffdd3-8f1a-4bae-a6d1-b573934cc09d" />

This PR fixes the backend and frontend issues preventing the Educator Task Submissions page from working correctly and completes missing logic requested by Harshitha & Sphurthy.
The previous PR (#4207 frontend, #1799 backend) implemented the UI but failed to load or filter student submissions due to missing studentId and incorrect filtering logic.
This PR resolves those issues and adds improved functionality.

## Related PRS (if any):
This frontend PR is related to the backend [PR#1934](https://github.com/OneCommunityGlobal/HGNRest/pull/1934)

## Main changes explained:

- Updated frontend to use new studentId correctly
- Fixed filter logic so:
 	   "All Submissions" → returns all
           "Pending Review" → only pending
 	   "Graded" → only graded
 	       Before: always returned full list
 	       After: correct filtering based on backend data
- Ensured data groups by:
 	lessonPlanId → classes
 	taskName → task groups inside each class
- Verified UI components load successfully after backend fix
 	TaskSubmissionsPage
 	SubmissionCard
 	Tabs, dropdown, tooltips, card states

 



## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. Clear site data/cache
4. Login and Navigation:
Login with an educator account
Navigate to /educationportal/educator/task-submissions
Verify the page loads with submission data

Verify:
Tabs load class names
Tasks expand correctly
Cards show:
pending → info tooltip
graded → badge
late → red styling
Filtering toggles results properly
No more infinite loading or 404/401 issues

## Screenshots or videos of changes:

## Note:
Include the information the reviewers need to know.

linh2020 committed:
Confirm educator task submissions UI with late submission tagging

Anusha-Gali reviewed (changes requested):
Hi Linh,

I have reviewed your PR locally and as you can see in the below video, the huge unnecessary data i.e neither graded or pending review is causing lag and is breaking the mobile responsiveness.
<img width="1272" height="755" alt="Screenshot 2025-12-24 at 5 20 49 PM" src="https://github.com/user-attachments/assets/6fc05cd6-80a2-451a-ae9b-1af03f8f38cc" />

https://github.com/user-attachments/assets/0d115f11-edc1-4316-8917-e3728708adc1

linh2020 committed:
fix(education-portal): fix lag, sub-task toast spam, and mobile responsiveness

- Suppress 'Failed to fetch sub-tasks' toast on 404 (demo/mock task IDs
  that don't exist in the DB no longer flood the UI with error toasts)
- Add Active / Pending Review / Completed filter tabs so graded and
  completed tasks are hidden by default, eliminating unnecessary renders
- Lazy-load intermediate tasks only for the currently visible (filtered)
  tasks; cache results to avoid redundant API calls on tab switch
- Replace overflow nav bar with a hamburger menu on mobile (≤768px)
- Fix card grid min-width and single-column collapse on mobile
- Remove fixed min-height from TaskCard; fix action-button layout at
  small breakpoints (480px) for both card and list views

linh2020 committed:
Merge origin/development into linh_educator_taskSubmissions_frontend_uiFilter

Commit:
fix(merge): resolve conflicts with development branch

Commit:
fix(lint): remove extra blank line in StudentDashboard

Commit:
fix(dark-mode): add dark mode styles for educator task submissions page

Commit:
fix(sonarcloud): handle unused catch binding and remove unused AssignAtoms import

Commit:
fix(sonarcloud): use Set for status arrays, fix catch bindings, extract nested ternary

kzou55 reviewed (changes requested):
I ran and tested the current branch(and associated backend branch) locally. I can verify that the tabs load class names, the tasks expand correctly. And that the cards show the correct components/information. And that the filter toggles results correctly. 


Testing in Light Mode

https://github.com/user-attachments/assets/81388448-6bee-452e-a43e-dc4ad8fc6b20


Testing in Dark Mode

https://github.com/user-attachments/assets/58b8fe89-3a7c-40af-a947-2728f4077699


Issues/Concerns
- So the filter works as stated in the description and  how to test section, however, is there a reason why filtering by "Submission Pending" includes Late Submissions? And that there isn't an option in the submission filter to be only late submissions?
 
  <img width="1900" height="1030" alt="Image" src="https://github.com/user-attachments/assets/ab9daf4c-94d1-401b-9586-8dab4b8bfdba" />

  <img width="290" height="169" alt="Image" src="https://github.com/user-attachments/assets/e368c6ba-b5c7-416a-b1b4-a844e0b13fb8" />

AmaanSyed09 reviewed (dismissed):
Tested frontend PR #4486 together with backend PR #1934.

Environment:

1. Checked out frontend branch: linh_educator_taskSubmissions_frontend_uiFilter

2. Checked out backend branch: linh_educator_taskSubmissions_backend_lateDetection

3. Ran frontend with npm run start:local

4. Ran backend with npm run dev

5. Cleared cache/site data before testing

Verified:

1. Educator Task Submissions page loads successfully.

2. Verified the page renders correctly in both Light Mode and Dark Mode.

3. Verified course tabs load and switch correctly.

4. Verified task groups expand and collapse correctly.

5. Verified submission cards render correctly.

6. Verified late submissions display the expected red styling.

7. Verified late submission tooltip displays the correct submission timestamp.

8. Verified the All Submissions, Submissions Pending, and Submissions Received filters work as expected.

9. Verified no console errors were encountered during testing (only non-blocking warnings).

Issues Found:

None.

<img width="468" height="259" alt="Image" src="https://github.com/user-attachments/assets/d902ca1b-6862-4a7b-a365-e75c5ce5d123" />

<img width="468" height="257" alt="Image" src="https://github.com/user-attachments/assets/b19efc33-ae55-43a5-83eb-7a391146e8c5" />

<img width="468" height="188" alt="Image" src="https://github.com/user-attachments/assets/02df8041-d5e7-45a8-842b-bd07b53fc966" />

<img width="468" height="231" alt="Image" src="https://github.com/user-attachments/assets/ca6fa5b8-bb5d-454e-919b-f56bbae030a3" />

<img width="468" height="229" alt="Image" src="https://github.com/user-attachments/assets/df1a63a4-7391-4bb8-ae19-59724af18dc3" />

<img width="468" height="352" alt="Image" src="https://github.com/user-attachments/assets/066d50d6-4830-4e7b-80ec-133f470e70d9" />

<img width="468" height="409" alt="Image" src="https://github.com/user-attachments/assets/28ba5620-8c59-435f-8ffc-83dcf401cc24" />

Commit:
fix: match filter dropdown to spec, show late status on graded cards, fix card overflow

DeepighaJ reviewed (changes requested):
Test the PR and the task are filtered as per the filter selection. 
I noticed the pending tasks doesnot have the tooltop info icon under pending Submision filter but icon is available in All submission level.
Otherwise all functionalities are as expected.

<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/109a23e0-f5f9-4265-ab97-1b2665ae1e1d" />
<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/64533381-aeaf-435f-9787-f48d9f5e430a" />
<img width="1600" height="852" alt="Image" src="https://github.com/user-attachments/assets/db4f1390-20a8-42d9-a83f-95f22d941124" />

Commit:
chore: merge development, resolve conflicts, add submission review page

sawantgayatri19 reviewed (changes requested):
Hello @amaresh2001,

Thanks for implementing the recent updates, including the dark mode styles, the pending submission info icon / "Not yet submitted" tooltips, and the new educator review page (/educationportal/educator/review/:submissionId).

To help us get this merged cleanly, could you please clarify or touch base on @kzou55's question regarding whether a separate filter option is needed specifically for late submissions (since they currently show under "Submission Pending")? Once that's clarified, we should be all set!

RuthwikHM reviewed (changes requested):
Tested this out locally with https://github.com/OneCommunityGlobal/HGNRest/pull/1934. Checked whether filtering itself works fine by status alongside the tabs having the right class names, expanding and closing correctly, the cards showing the right badges for graded, late submissions, pending. Found no console errors either. I did notice two bugs in this page. I've left a comment in the code detailing one of them. The other can be found below.

The scroll to top of page button seems to be broken. Not sure if this is from changes in this PR or elsewhere. But fixing this would be a nice improvement to user experience.
https://github.com/user-attachments/assets/5f344aba-1f55-4f61-be69-449ff3bb97d1

Commit:
chore: merge development, resolve routes.jsx conflict, keep review page route

Commit:
fix: preserve active tab/expanded state across filter changes, fix broken back-to-top button

Commit:
fix: resolve SonarCloud code smells - explicit button types, semantic button element, accessible roles, handled exceptions

Commit:
fix: keep course tabs stable across filters, restore view on back-nav, fix dark-mode status text color

ShreyasB02 reviewed (dismissed):
Hello,
Tested all of the following changes locally with [HGNRest/pull/1934](https://github.com/OneCommunityGlobal/HGNRest/pull/1934#pullrequestreview-4933701344 ) and verified they are working as expected:

* Verified the submissions page loads correctly with submission data.
* Confirmed class tabs load with the correct class names.
* Verified task groups expand and collapse correctly.
* Validated the updated filter logic:

  * **All Submissions** → displays all submissions.
  * **Pending Review** → displays only pending submissions.
  * **Graded** → displays only graded submissions.
* Tested pending and late submissions together to ensure their states do not conflict or mask each other.
* Verified submission UI states:

  * **Pending** → displays the info tooltip.
  * **Graded** → displays the graded badge.
  * **Late** → displays the correct red styling.
* Confirmed submissions are grouped correctly:

  * By `lessonPlanId` → into classes.
  * By `taskName` → into task groups within each class.
* Verified all UI components render successfully following the backend changes:

  * `TaskSubmissionsPage`
  * `SubmissionCard`
  * Tabs
  * Dropdowns
  * Tooltips
  * Card state indicators
* Verified the previously reported **scroll-to-top** issue has been resolved and is working correctly.

Everything above has been tested locally and is functioning as expected.

##### Approving these changes.
---

<img width="1470" height="956" alt="Screenshot 2026-08-13 at 11 01 03 PM" src="https://github.com/user-attachments/assets/58a4da49-0957-4467-a680-df749694bceb" />
<img width="1470" height="956" alt="Screenshot 2026-08-13 at 11 01 36 PM" src="https://github.com/user-attachments/assets/49da368f-8389-4cf4-b44b-1a63915cf9a0" />
<img width="1470" height="956" alt="Screenshot 2026-08-13 at 11 05 22 PM" src="https://github.com/user-attachments/assets/b8900e22-470a-4bf0-aa5d-00020ef7504b" />
<img width="1470" height="956" alt="Screenshot 2026-08-13 at 11 13 11 PM" src="https://github.com/user-attachments/assets/58365a01-1217-41fb-8267-05decd49d5af" />

RuthwikHM reviewed (dismissed):
Thanks for fixing the tab expansion and selection bugs!! LGTM!!

Adit0717 reviewed (changes requested):
Tested the PR locally with the specified frontend and backend branches. All of the functionalities work but I noted an issue in the Marks field. The field doesn't validate against the "Out of" value. For example, I was able to enter 123123123154 even though the max is 100.

<img width="674" height="892" alt="Image" src="https://github.com/user-attachments/assets/2fc8fa50-7423-4719-b91a-5982bcd7adfb" />

Commit:
fix: validate marks do not exceed out-of value in grade submission

amaresh2001 commented:
@Adit0717 Thanks for catching this! Fixed it; the Marks input was missing a max constraint, so it accepted anything.
Please test this when you get a chance!

kzou55 reviewed (approved):
Hi Amaresh, 

I ran and tested the PR locally alongside its associated backend. 

## Verified
- The page loads with submission data and tabs with class names

    <img width="1909" height="1026" alt="Image" src="https://github.com/user-attachments/assets/e90248f9-d589-4e31-a764-865d140177c0" />

    <img width="1900" height="1025" alt="Image" src="https://github.com/user-attachments/assets/a56849b6-b87d-403b-a3b9-6f43a5d3c5b6" />
- Tasks expand correctly
  - And the cards contains either the following: pending → info tooltip, graded → badge, late → red styling 

    <img width="1911" height="1021" alt="Image" src="https://github.com/user-attachments/assets/8f139688-fa1a-4824-87bf-b3885eefe4da" />

    <img width="1906" height="1024" alt="Image" src="https://github.com/user-attachments/assets/404882c3-fffa-4625-8099-0bb569020238" />

- Filter dropdown works

    <img width="1880" height="237" alt="Image" src="https://github.com/user-attachments/assets/20b00491-372d-41af-b79d-3f1dd88f811b" />

    <img width="1881" height="218" alt="Image" src="https://github.com/user-attachments/assets/b98f66ff-ccf9-41ed-8905-06be6167a13c" />

  - Pending Submission

     <img width="1910" height="1036" alt="Image" src="https://github.com/user-attachments/assets/b7805a0d-60ae-4c07-8b7b-69c01923dd72" />

    <img width="1915" height="1037" alt="Image" src="https://github.com/user-attachments/assets/ceb3ac9a-1fd9-440e-a9e4-419edd100a4b" />
- Graded 

    <img width="1899" height="901" alt="Image" src="https://github.com/user-attachments/assets/f28c82c7-cbf0-44ae-b051-3d98f9b99a3c" />

    <img width="1909" height="910" alt="Image" src="https://github.com/user-attachments/assets/dc7e9126-2573-4526-a635-34cd924045ea" />

- Late Submission

    <img width="1886" height="783" alt="Image" src="https://github.com/user-attachments/assets/01447d96-a091-42df-b9ed-122dafc3520e" />

    <img width="1898" height="767" alt="Image" src="https://github.com/user-attachments/assets/cbd93905-3fd8-4685-83b3-60b2dc65ef68" />

- Marks cannot exceed out of value

    <img width="838" height="866" alt="Image" src="https://github.com/user-attachments/assets/dfcffd63-8763-4f42-9cae-f3dc678b3897" />

    <img width="784" height="873" alt="Image" src="https://github.com/user-attachments/assets/c99a6376-03b6-45b8-aedd-2f81a94d3ce2" />

one-community commented:
Thank you all, merging!

iAbhi001 commented:
Here is a standard, comprehensive PR approval template you can post as your GitHub review comment:

---

Tested backend PR **#1934** locally alongside frontend PR **#4486**. Everything is working as expected.

**Testing & Verification Summary:**

* **Integration & Startup:** Backend connected to MongoDB without errors, and the frontend app rendered the submissions view cleanly at `/educationportal/educator/task-submissions`.
* **Data Flow & UI Display:** Submissions loaded dynamically with complete record details (student name, task, type/link, submission date, and status).
* **Filters:** Verified filtering logic by course, submission status, and student. Data updates dynamically and accurately.
* **Late & Overdue Detection:**
* Tasks completed after the due date correctly show the **Late** tag and valid time offset data.
* Past-due unsubmitted tasks properly reflect **Overdue** status under the pending filter.


* **API & Security:** Confirmed unauthenticated requests correctly return `401 Unauthorized`, and invalid query/payload scenarios return expected validation errors.
* **Automated Tests:** All local test suites passed with no regressions.

The changes look solid. Approving!

<img width="1470" height="956" alt="image" src="https://github.com/user-attachments/assets/2bfdd77f-8781-4065-baad-c5ffed7cd2f3" />


<img width="1470" height="956" alt="image" src="https://github.com/user-attachments/assets/656db6d2-b697-4ef7-849f-c0293f364919" />
